### PR TITLE
HAAR-3389 Validation rules for new AuditFilterDto

### PR DESCRIPTION
**Pull request by ma226860**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+class DigitalServicesAuditFilterValidator : ConstraintValidator<ValidDigitalServicesAuditFilter, DigitalServicesAuditFilterDto> {
+
+  override fun isValid(dto: DigitalServicesAuditFilterDto?, context: ConstraintValidatorContext): Boolean {
+    if (dto == null) return false
+    context.disableDefaultConstraintViolation()
+    val now = Instant.now()
+    var isValid = true
+
+    if (dto.startDateTime == null && dto.endDateTime == null) {
```
Comment (mpbostock) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:15]: If one or the other of these needs to be provided, should this be a check to see if start is null when end is not and another check for vice versa?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    if (dto.who == null && dto.subjectId == null && dto.subjectType == null) {
+      isValid = false
+      context.buildConstraintViolationWithTemplate("If who is null then subjectId and subjectType must be populated")
+        .addPropertyNode("who")
+        .addConstraintViolation()
+    }
+
+    if (dto.endDateTime != null && dto.endDateTime!!.isAfter(now)) {
```
Comment (mpbostock) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:33]: Because we've already checked for null in the first part of the if, we don't need the `!!` in the second as Kotlin knows that it's not null here.

We could also use this too:
```kotlin
dto.endDateTime?.isAfter(now) == true
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      isValid = false
+      context.buildConstraintViolationWithTemplate("endDateTime must not be in the future")
+        .addPropertyNode("endDateTime")
+        .addConstraintViolation()
+    }
+
+    if (dto.startDateTime != null && dto.startDateTime!!.isAfter(now)) {
```
Comment (mpbostock) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:40]: Same comment as endDateTime

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      isValid = false
+      context.buildConstraintViolationWithTemplate("startDateTime must not be in the future")
+        .addPropertyNode("startDateTime")
+        .addConstraintViolation()
+    }
+
+    if (dto.startDateTime != null && dto.endDateTime != null && dto.startDateTime!!.isAfter(dto.endDateTime)) {
```
Comment (mpbostock) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:47]: We also don't need the `!!` in the last part of the if here

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        ),
+      )
+
+    private fun invalidBaseAuditFilterDto() =
+      Stream.of(
+        Arguments.of(
+          DigitalServicesAuditFilterDto(),
+          mapOf(
+            "startDateTime" to "startDateTime must be provided if endDateTime is null",
+            "endDateTime" to "endDateTime must be provided if startDateTime is null",
```
Comment (mpbostock) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidatorTest.kt:74]: Can we add a test to check if one or the other of `startDateTime` or `endDateTime` is null?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      isValid = false
+      context.buildConstraintViolationWithTemplate("startDateTime must be before endDateTime")
+        .addPropertyNode("startDateTime")
+        .addConstraintViolation()
+    }
+
+    if ((dto.subjectId != null && dto.subjectType == null) || (dto.subjectId == null && dto.subjectType != null)) {
```
Comment (mpbostock) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:54]: Are the number of ifs here likely to increase? There's repetition in each of them, each one has basically a predicate on a dto that if fails sets the isValid to be false and builds a constraint violation, I'm wondering if each rule could be put in an Enum class or something that each one knows it's predicate and adds the constraint on if it fails, this method could then basically check if any in the list of enums has failed to work out isValid.

If there's going to be different Dtos with common fields such as dateTime and those need to be validated in the same way (i.e. not in the future etc.) it could be that each field potentially has a validator that we just re-use and annotate each field with their own validator. Are there any validators out of the box that would support a dateTime field??

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt (hunk 1 above)
Comment (mpbostock) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:15]: Actually I think I've misread this one, if one is null the other has to be provided so this is the only check you can do.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidatorTest.kt (hunk 5 above)
Comment (mpbostock) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidatorTest.kt:74]: I've misread this, apologies, this comment no longer applies

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt (hunk 6 above)
Comment (ma226860) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/model/DigitalServicesAuditFilterValidator.kt:54]: As discussed, can do as a refactor when we need to do something similar again
